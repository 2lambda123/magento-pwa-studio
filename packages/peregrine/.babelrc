{
  "plugins": [
    "transform-object-rest-spread",
    "transform-class-properties",
    ["transform-react-jsx", { "pragma": "createElement" }],
    ["transform-runtime", {
      "helpers": true,
      "polyfill": false, // polyfills will be handled by preset-env
      "regenerator": false
    }]
  ],
  "presets": [
    ["env", {
      "targets": {
        "browsers": ["> 5%"]
      },
      "modules": false
    }]
  ],
  "env": {
    "test": {
      "plugins": [
        "transform-es2015-modules-commonjs",
        "transform-object-rest-spread",
        "transform-class-properties",
        ["transform-react-jsx", { "pragma": "createElement" }]
      ]
    }
  }
}
