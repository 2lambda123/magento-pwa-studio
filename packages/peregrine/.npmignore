.eslintrc.js
.storybook
/lib/**/__{docs,helpers,mocks,stories,tests}__/**
/scripts
