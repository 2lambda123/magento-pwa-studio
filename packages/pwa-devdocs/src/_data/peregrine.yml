title: Peregrine
entries:
  - label: Overview
    url: /peregrine/

  - label: Storybook UI testing
    url: /peregrine/testing/storybook/

  - label: Utility modules
    entries:

      - label: REST API client
        url: /peregrine/reference/rest-api-client/

  - label: React components
    entries:

      - label: ContainerChild
        url: /peregrine/reference/container-child/

      - label: Item
        url: /peregrine/reference/item/

      - label: Items
        url: /peregrine/reference/items/

      - label: List
        url: /peregrine/reference/list/

      - label: Router
        url: /peregrine/reference/router/
