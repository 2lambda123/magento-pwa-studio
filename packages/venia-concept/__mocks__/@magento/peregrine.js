import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/src');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useDocumentListener = jest.fn(peregrine.useDocumentListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);

// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useDocumentListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam
};
