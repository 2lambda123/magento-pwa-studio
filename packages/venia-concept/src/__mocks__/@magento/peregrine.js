import { createElement } from 'react';

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
export const Price = () => <div />;
