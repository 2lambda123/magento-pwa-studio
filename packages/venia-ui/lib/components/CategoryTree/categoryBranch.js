import React, { useCallback } from 'react';
import { func, number, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import defaultClasses from './categoryBranch.css';

const Branch = props => {
    const { category, setCategoryId } = props;
    const { id, include_in_menu, name } = category;
    const classes = mergeClasses(defaultClasses, props.classes);

    const handleClick = useCallback(() => {
        setCategoryId(id);
    }, [id, setCategoryId]);

    // `include_in_menu` is undefined when Magento <= 2.3.1
    if (include_in_menu === 0) {
        return null;
    }

    return (
        <li className={classes.root}>
            <button
                className={classes.target}
                type="button"
                onClick={handleClick}
            >
                <span className={classes.text}>{name}</span>
            </button>
        </li>
    );
};

export default Branch;

Branch.propTypes = {
    category: shape({
        id: number.isRequired,
        include_in_menu: number,
        name: string.isRequired
    }).isRequired,
    classes: shape({
        root: string,
        target: string,
        text: string
    }),
    setCategoryId: func.isRequired
};
