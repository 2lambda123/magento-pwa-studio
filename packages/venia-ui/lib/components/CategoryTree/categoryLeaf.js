import React, { useCallback } from 'react';
import { func, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import { Link, resourceUrl } from '../../drivers';
import defaultClasses from './categoryLeaf.css';

const suffix = '.html';

const Leaf = props => {
    const { category, onNavigate } = props;
    const { name, url_path } = category;
    const classes = mergeClasses(defaultClasses, props.classes);

    const handleClick = useCallback(() => {
        onNavigate();
    }, [onNavigate]);

    return (
        <li className={classes.root}>
            <Link
                className={classes.target}
                to={resourceUrl(`/${url_path}${suffix}`)}
                onClick={handleClick}
            >
                <span className={classes.text}>{name}</span>
            </Link>
        </li>
    );
};

export default Leaf;

Leaf.propTypes = {
    category: shape({
        name: string.isRequired,
        url_path: string.isRequired
    }).isRequired,
    classes: shape({
        root: string,
        target: string,
        text: string
    }),
    onNavigate: func.isRequired
};
