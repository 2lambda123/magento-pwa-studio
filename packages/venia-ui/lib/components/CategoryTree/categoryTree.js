import React, { useEffect } from 'react';
import { func, number, objectOf, shape, string } from 'prop-types';
import { useQuery } from '@magento/peregrine';

import { mergeClasses } from '../../classify';
import MENU_QUERY from '../../queries/getNavigationMenu.graphql';
import Branch from './categoryBranch';
import Leaf from './categoryLeaf';
import defaultClasses from './categoryTree.css';

const Tree = props => {
    const {
        categories,
        categoryId,
        onNavigate,
        setCategoryId,
        updateCategories
    } = props;

    const classes = mergeClasses(defaultClasses, props.classes);
    const [queryResult, queryApi] = useQuery(MENU_QUERY);
    const { data } = queryResult;
    const { runQuery } = queryApi;

    // fetch categories
    useEffect(() => {
        if (categoryId != null) {
            runQuery({ variables: { id: categoryId } });
        }
    }, [categoryId, runQuery]);

    // update redux with fetched categories
    useEffect(() => {
        if (data && data.category) {
            updateCategories(data.category);
        }
    }, [data, updateCategories]);

    const rootCategory = categories[categoryId];
    const { children } = rootCategory || {};

    // for each child category, render a direct link if it has no children
    // otherwise render a branch
    const branches = Array.from((rootCategory && children) || [], id => {
        const category = categories[id];
        const isLeaf = category.children_count === '0';

        return isLeaf ? (
            <Leaf key={id} category={category} onNavigate={onNavigate} />
        ) : (
            <Branch
                key={id}
                category={category}
                setCategoryId={setCategoryId}
            />
        );
    });

    return (
        <div className={classes.root}>
            <ul className={classes.tree}>{branches}</ul>
        </div>
    );
};

export default Tree;

Tree.propTypes = {
    categories: objectOf(
        shape({
            id: number.isRequired,
            name: string
        })
    ),
    categoryId: number.isRequired,
    classes: shape({
        root: string,
        tree: string
    }),
    onNavigate: func.isRequired,
    setCategoryId: func.isRequired,
    updateCategories: func.isRequired
};
