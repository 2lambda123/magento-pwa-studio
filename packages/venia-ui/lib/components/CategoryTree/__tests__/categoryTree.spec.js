import React from 'react';
import { act } from 'react-test-renderer';
import { createTestInstance, useQuery } from '@magento/peregrine';

import CategoryTree from '../categoryTree';

jest.mock('@magento/peregrine');
jest.mock('../../../classify');
jest.mock('../categoryBranch', () => () => <i />);
jest.mock('../categoryLeaf', () => () => <i />);

const queryState = {
    data: null,
    error: null,
    loading: false
};
const queryApi = {
    resetState: jest.fn(),
    runQuery: jest.fn(),
    setLoading: jest.fn()
};

useQuery.mockImplementation(() => [queryState, queryApi]);

const props = {
    categories: {},
    categoryId: 1,
    onNavigate: jest.fn(),
    setCategoryId: jest.fn(),
    updateCategories: jest.fn()
};

const categories = {
    1: {
        children: [2, 4],
        children_count: 2,
        id: 1,
        name: 'One',
        path: '1'
    },
    2: {
        id: 2,
        include_in_menu: 1,
        name: 'Two',
        parentId: 1,
        path: '1/2',
        productImagePreview: {
            items: [{ small_image: 'media/img-2.jpg' }]
        }
    },
    3: {
        id: 3,
        include_in_menu: 1,
        name: 'Three',
        parentId: 2,
        path: '1/2/3',
        productImagePreview: {
            items: [{ small_image: 'media/img-3.jpg' }]
        }
    },
    4: {
        id: 4,
        include_in_menu: 1,
        name: 'Four',
        parentId: 1,
        path: '1/4',
        productImagePreview: {
            items: [{ small_image: 'media/img-4.jpg' }]
        }
    }
};

test('renders correctly without data', async () => {
    const instance = createTestInstance(<CategoryTree {...props} />);

    expect(instance.toJSON()).toMatchSnapshot();
});

test('calls runQuery on mount', () => {
    const { runQuery } = queryApi;
    createTestInstance(<CategoryTree {...props} />);

    act(() => {});

    expect(runQuery).toHaveBeenCalledTimes(1);
    expect(runQuery).toHaveBeenNthCalledWith(1, {
        variables: {
            id: props.categoryId
        }
    });
});

test('calls runQuery when categoryId changes', () => {
    const { runQuery } = queryApi;
    const instance = createTestInstance(<CategoryTree {...props} />);

    act(() => {
        instance.update(<CategoryTree {...props} categoryId={2} />);
    });

    expect(runQuery).toHaveBeenCalledTimes(2);
    expect(runQuery).toHaveBeenNthCalledWith(2, {
        variables: {
            id: 2
        }
    });
});

test('avoids calling runQuery without a category id', () => {
    const { runQuery } = queryApi;
    createTestInstance(<CategoryTree {...props} categoryId={null} />);

    act(() => {});

    expect(runQuery).not.toHaveBeenCalled();
});

test('avoids calling updateCategories without data', () => {
    const { updateCategories } = props;
    createTestInstance(<CategoryTree {...props} />);

    act(() => {});

    expect(updateCategories).not.toHaveBeenCalled();
});

test('renders correctly with data', () => {
    const instance = createTestInstance(
        <CategoryTree {...props} categories={categories} />
    );

    expect(instance.toJSON()).toMatchSnapshot();
});

test('calls updateCategories when data changes', () => {
    const { updateCategories } = props;
    const category = categories[1];
    const data = {
        category: {
            ...category,
            children: Array.from(category.children, id => categories[id])
        }
    };

    const nextQueryState = { ...queryState, data };

    useQuery.mockImplementationOnce(() => [nextQueryState, queryApi]);

    createTestInstance(<CategoryTree {...props} />);

    act(() => {});

    expect(updateCategories).toHaveBeenCalledTimes(1);
    expect(updateCategories).toHaveBeenNthCalledWith(1, data.category);
});

// test('child node correctly sets new root and parent ids', async () => {
//     const rootNodeId = 1;
//     const currentId = 1;
//     const setCurrentPath = jest.fn();

//     const { root } = TestRenderer.create(
//         <MockedProvider mocks={mocks} addTypename={false}>
//             <MemoryRouter>
//                 <CategoryTree
//                     rootNodeId={rootNodeId}
//                     currentId={currentId}
//                     updateRootNodeId={setCurrentPath}
//                     classes={classes}
//                 />
//             </MemoryRouter>
//         </MockedProvider>
//     );

//     await waitForExpect(() => {
//         const child = root.findByProps({ path: '1/3' });
//         const { onDive, path } = child.props;

//         onDive(path);

//         expect(setCurrentPath).toHaveBeenLastCalledWith(path);
//     });
// });
