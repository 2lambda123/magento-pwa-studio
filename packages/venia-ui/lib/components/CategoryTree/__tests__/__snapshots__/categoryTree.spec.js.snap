// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with data 1`] = `
<div
  className="root"
>
  <ul
    className="tree"
  >
    <i />
    <i />
  </ul>
</div>
`;

exports[`renders correctly without data 1`] = `
<div
  className="root"
>
  <ul
    className="tree"
  />
</div>
`;
